internal/gitea/repo.go

```diff
@@ -0,0 +1,190 @@
+		})
+	}
+
+	return nil
+}
+
+// processCommitsSinceRelease calculates the number of commits that have occurred on the default
+// branch of the repository since the last release, and populates the information in r.Details.
+func (r *Repository) processCommitsSinceRelease() error {
+	// This does not currently handle monorepos - see note about releases in
+	// `processReleases`. If there are releases in Gitea with a monorepo, then
+	// this function maybe needs to restrict the commits to ones where the tree
+	// overlaps - but maybe we actually want to know how many commits overall,
+	// because we want to count things like common code being adjusted.
+
+	// Add the commit delta between last release and default branch.
+	latestRelease := r.Details.Releases[len(r.Details.Releases)-1]
+	opts := gitea.ListCommitOptions{ListOptions: gitea.ListOptions{PageSize: giteaReleasesPerRepo}, SHA: latestRelease.Version, Path: ""}
+	commits, _, err := r.client.ListRepoCommits(r.org, r.Details.Name, opts)
+	if err != nil {
+		return errors.New("error getting commits since release")
+	}
+
+	r.Details.NewCommits = len(commits)
+
+	return nil
+}
+
+// processCommits fetches the latest 3 commits to a repository and populates them into the repo
+// struct in the case that there are no releases identified.
+func (r *Repository) processCommits() error {
+	// This does not currently handle monorepos. It's not clear which commits
+	// should be counted in this case - only ones that have a tree that overlaps
+	// with the subfolder? All commits, as now? If there's common code, then
+	// all commits is probably truest to the single-repo meaning, but if there
+	// are commits that are only in a separate charm, then those probably should
+	// not be included.
+
+	opts := gitea.ListCommitOptions{ListOptions: gitea.ListOptions{PageSize: giteaReleasesPerRepo}, SHA: r.defaultBranch, Path: ""}
+
+	// If there are no releases, get the latest commit instead.
+	commits, _, err := r.client.ListRepoCommits(r.org, r.Details.Name, opts)
+	if err != nil {
+		return errors.New("error listing commits for repository")
+	}
+
+	// Iterate over the commits and append them to r.Details.Commits
+	for _, commit := range commits {
+		// Some commits don't have an author.
+		var name string
+		if commit.Author == nil {
+			name = ""
+		} else {
+			name = commit.Author.FullName
+		}
+		r.Details.Commits = append(r.Details.Commits, &repos.Commit{
+			Sha:       commit.CommitMeta.SHA,
+			Author:    name,
+			Timestamp: commit.CommitMeta.Created.Unix(),
+			Message:   renderReleaseBody(commit.RepoCommit.Message, r),
+			URL:       commit.HTMLURL,
+		})
+	}
+
+	return nil
+}
+
+// renderReleaseBody transforms a Markdown string from a Gitea Release into HTML.
+func renderReleaseBody(body string, repo *Repository) string {
+	// Render the Markdown to HTML.
+	md := []byte(body)
+	normalised := markdown.NormalizeNewlines(md)
```

Comment:
No real need for this to be a named variable:

```suggestion
	normalised := markdown.NormalizeNewlines([]byte(body))
```